lib/std/cache_hash.zig

```diff
@@ -0,0 +1,507 @@
+        self.alloc.free(contents);
+    }
+
+    /// Returns a base64 encoded hash of the inputs.
+    pub fn final(self: *@This()) [BASE64_DIGEST_LEN]u8 {
+        debug.assert(self.manifest_file != null);
+
+        // We don't close the manifest file yet, because we want to
+        // keep it locked until the API user is done using it.
+        // We also don't write out the manifest yet, because until
+        // cache_release is called we still might be working on creating
+        // the artifacts to cache.
+
+        var bin_digest: [BIN_DIGEST_LEN]u8 = undefined;
+        self.blake3.final(&bin_digest);
+
+        var out_digest: [BASE64_DIGEST_LEN]u8 = undefined;
+        base64_encoder.encode(&out_digest, &bin_digest);
+
+        return out_digest;
+    }
+
+    pub fn write_manifest(self: *@This()) !void {
+        debug.assert(self.manifest_file != null);
+
+        var encoded_digest: [BASE64_DIGEST_LEN]u8 = undefined;
+        var contents = ArrayList(u8).init(self.alloc);
+        var outStream = contents.outStream();
+        defer contents.deinit();
+
+        for (self.files.items) |file| {
+            base64_encoder.encode(encoded_digest[0..], &file.bin_digest);
+            try outStream.print("{} {} {} {}\n", .{ file.stat.inode, file.stat.mtime, encoded_digest[0..], file.path });
+        }
+
+        try self.manifest_file.?.seekTo(0);
+        try self.manifest_file.?.writeAll(contents.items);
+    }
+
+    /// Releases the manifest file and frees any memory the CacheHash was using.
+    /// `CacheHash.hit` must be called first.
+    ///
+    /// Will also attempt to write to the manifest file if the manifest is dirty.
+    /// Writing to the manifest file is the only way that this file can return an
+    /// error.
+    pub fn release(self: *@This()) !void {
+        debug.assert(self.manifest_file != null);
+
+        if (self.manifest_dirty) {
+            try self.write_manifest();
+        }
+
+        self.manifest_file.?.close();
+
+        for (self.files.items) |*file| {
+            file.deinit(self.alloc);
+        }
+        self.files.deinit();
+        self.manifest_dir.close();
+    }
+};
+
+/// Hash the file, and return the contents as an array
+fn hash_file(alloc: *Allocator, bin_digest: []u8, handle: *const fs.File) ![]u8 {
+    var blake3 = Blake3.init();
+
+    const contents = try handle.inStream().readAllAlloc(alloc, 64 * 1024);
+
+    blake3.update(contents);
+
+    blake3.final(bin_digest);
+
+    return contents;
+}
+
+/// If the wall clock time, rounded to the same precision as the
+/// mtime, is equal to the mtime, then we cannot rely on this mtime
+/// yet. We will instead save an mtime value that indicates the hash
+/// must be unconditionally computed.
+fn is_problematic_timestamp(file_mtime_ns: i64) bool {
+    const now_ms = time.milliTimestamp();
+    const file_mtime_ms = @divFloor(file_mtime_ns, time.millisecond);
+    return now_ms == file_mtime_ms;
+}
```

Comment:
I don't think this has the same behavior as the c++ code:

```c++
static bool is_problematic_timestamp(const OsTimeStamp *fs_clock) {
    OsTimeStamp wall_clock = os_timestamp_calendar();
    // First make all the least significant zero bits in the fs_clock, also zero bits in the wall clock.
    if (fs_clock->nsec == 0) {
        wall_clock.nsec = 0;
        if (fs_clock->sec == 0) {
            wall_clock.sec = 0;
        } else {
            wall_clock.sec &= (-1ull) << ctzll(fs_clock->sec);
        }
    } else {
        wall_clock.nsec &= (-1ull) << ctzll(fs_clock->nsec);
    }
    return wall_clock.nsec == fs_clock->nsec && wall_clock.sec == fs_clock->sec;
}
```

`std.time.milliTimeStamp` is not quite what we want here. It looks like `std.time.milliTimeStamp` is too high level, and `std.time` does not actually offer what we need: a cross-platform high resolution calendar timestamp. This will require a new function added to `std.time`. I believe that `std.time.milliTimeStamp` could then be reworked to call this other function, and do a small amount of logic on top of it.
`os_timestamp_calendar` corresponds to